src/CounterInterceptor/MyCounterInterceptor.cs

```diff
@@ -0,0 +1,137 @@
+namespace TemporalioSamples.CounterInterceptor;
+
+using System.Collections.Concurrent;
+using Temporalio.Activities;
+using Temporalio.Client;
+using Temporalio.Client.Interceptors;
+using Temporalio.Worker.Interceptors;
+using Temporalio.Workflows;
+
+public class MyCounterInterceptor : IClientInterceptor, IWorkerInterceptor
+{
+    private ConcurrentDictionary<string, Counts> counts = new();
+
+    public ConcurrentDictionary<string, Counts> Counts => counts;
+
+    public string WorkerInfo() =>
+        string.Join(
+                "\n",
+                counts.Select(kvp => $"** Workflow ID: {kvp.Key} {kvp.Value.WorkflowInfo()}"));
+
+    public string ClientInfo() =>
+        string.Join(
+                "\n",
+                counts.Select(kvp => $"** Workflow ID: {kvp.Key} {kvp.Value.ClientInfo()}"));
+
+    public ClientOutboundInterceptor InterceptClient(ClientOutboundInterceptor nextInterceptor) =>
+        new ClientOutbound(this, nextInterceptor);
+
+    public WorkflowInboundInterceptor InterceptWorkflow(WorkflowInboundInterceptor nextInterceptor) =>
+        new WorkflowInbound(this, nextInterceptor);
+
+    public ActivityInboundInterceptor InterceptActivity(ActivityInboundInterceptor nextInterceptor) =>
+        new ActivityInbound(this, nextInterceptor);
+
+    private sealed class ClientOutbound : ClientOutboundInterceptor
+    {
+        private MyCounterInterceptor root;
+
+        public ClientOutbound(MyCounterInterceptor root, ClientOutboundInterceptor next)
+            : base(next) => this.root = root;
+
+        public override Task<WorkflowHandle<TWorkflow, TResult>> StartWorkflowAsync<TWorkflow, TResult>(
+            StartWorkflowInput input)
+        {
+            var id = input.Options.Id ?? "None";
+            // Need to add an empty record if none exists
+            // we don't care if it doesn't add it as we will
+            // still increment the current value.
+            root.counts.TryAdd(id, new Counts());
+            Interlocked.Increment(ref root.counts[id].ClientExecutions);
+            return base.StartWorkflowAsync<TWorkflow, TResult>(input);
+        }
+
+        public override Task SignalWorkflowAsync(SignalWorkflowInput input)
+        {
+            var id = input.Id ?? "None";
```

Comment:
`input.Id` can never be null on this call (and some others), so `??` should be removed